composer/2022_airflow_summit/data_analytics_process_expansion.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+import sys
+
+from py4j.protocol import Py4JJavaError
+from pyspark.sql import SparkSession
+from pyspark.sql.functions import avg, lit, sum, udf
+from pyspark.sql.types import FloatType, IntegerType, StructType
+
+# BQ_DESTINATION_DATASET_NAME = "expansion_project"
+# BQ_DESTINATION_TABLE_NAME = "ghcnd_stations_joined"
+# BQ_NORMALIZED_TABLE_NAME = "ghcnd_stations_normalized"
+# BQ_PRCP_MEAN_TABLE_NAME = "ghcnd_stations_prcp_mean"
+# BQ_SNOW_MEAN_TABLE_NAME = "ghcnd_stations_snow_mean"
+# BQ_PHX_PRCP_TABLE_NAME = "phx_annual_prcp"
+# BQ_PHX_SNOW_TABLE_NAME = "phx_annual_snow"
+
+if __name__ == "__main__":
+    BUCKET_NAME = sys.argv[1]
+    READ_TABLE = sys.argv[2]
+    DF_WRITE_TABLE = sys.argv[3]
+    PRCP_MEAN_WRITE_TABLE = sys.argv[4]
+    SNOW_MEAN_WRITE_TABLE = sys.argv[5]
+    PHX_PRCP_WRITE_TABLE = sys.argv[6]
+    PHX_SNOW_WRITE_TABLE = sys.argv[7]
+
+    # BUCKET_NAME = "workshop_example_bucket"
+    # READ_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_DESTINATION_TABLE_NAME}"
+    # DF_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_NORMALIZED_TABLE_NAME}"
+    # PRCP_MEAN_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_PRCP_MEAN_TABLE_NAME}"
+    # SNOW_MEAN_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_SNOW_MEAN_TABLE_NAME}"
+    # PHX_PRCP_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_PHX_PRCP_TABLE_NAME}"
+    # PHX_SNOW_WRITE_TABLE = f"{BQ_DESTINATION_DATASET_NAME}.{BQ_PHX_SNOW_TABLE_NAME}"
+
+    # Create a SparkSession, viewable via the Spark UI
+    spark = SparkSession.builder.appName("data_processing").getOrCreate()
+    # Load data into dataframe if READ_TABLE exists
+    try:
+        df = spark.read.format("bigquery").load(READ_TABLE)
+    except Py4JJavaError as e:
+        raise Exception(f"Error reading {READ_TABLE}") from e
+
+    # filter out non-west states of the US
+    df = df.where((df.STATE == 'WA') | (df.STATE == 'OR') | (df.STATE == 'ID')
+                  | (df.STATE == 'MT') | (df.STATE == 'WY') | (df.STATE == 'CO')
+                  | (df.STATE == 'NM') | (df.STATE == 'AZ') | (df.STATE == 'UT')
+                  | (df.STATE == 'NV') | (df.STATE == 'CA'))
```

Comment:
Will this not work if you move all of the states to a Python list?
```
western_states = ['WA', 'OR',...]
df = df.where(df.STATE not in western_states)
```

In either case, please organize the states alphabetically.

Reply:
Will do!

Reply:
It seems I can't move all the states to a Python list without getting an error. Here's the error message:
`File "/tmp/636e523659184e35bcd76cf35e7e0935/data_analytics_process_expansion.py", line 58, in <module>
    df = df.where((df.STATE in western_states))
  File "/usr/lib/spark/python/lib/pyspark.zip/pyspark/sql/column.py", line 905, in __nonzero__
ValueError: Cannot convert column into bool: please use '&' for 'and', '|' for 'or', '~' for 'not' when building DataFrame boolean expressions.`

Reply:
The problem is solved using the `isin()` func in PySpark:
`western_states = ['AZ', 'CA', 'CO', 'ID', 'MT', 'NM', 'NV', 'OR', 'UT', 'WA', 'WY']
df = df.where(df.STATE.isin(western_states))`